h2/src/main/org/h2/engine/Session.java

```diff
@@ -841,6 +818,9 @@ public void rollbackTo(Savepoint savepoint, boolean trimToSize) {
                 }
             }
         }
+        if(queryCache != null) {
+            queryCache.clear();
+        }
```

Comment:
why does the query cache need clearing?

Reply:
Because cache may have captured result based on uncommitted transaction’s data, which is not valid anymore, after rollback.

Reply:
Thanks, please add comment to that effect (so that nobody tries to "optimise" it away)

Reply:
I am talking about Query.lastResult content. Maybe it is too drastic and can be made more selective, but who optimizes rollbacks anyway?

Reply:
I don't care about optimising rollbacks(), that was just an example. I just care about accidental removal in the future because the person in question (who could easily be me) removes it because they don't understand it.

Reply:
Then the very first test will fail: testScript.sql:1621. I am not that smart to see it all through, just found it a hard way 🤓